<?xml version="1.0"?>
<launch>

  <!-- This launch assumes that you have already
       started you preferred RGB-D sensor and your IMU.
       TF between frame_id and the sensors should already be set too. -->

  <arg name="test_type"               default="gazebo"/>
  <arg name="frame_id"                default="base_footprint" />
  <arg name="output_pose_topic"       default="pose"/>
  <arg name="output_prefix"           default="/tmp/rtabmap"/>
  <arg name="output"                  default="log"/>

  <arg name="rgb_topic"               default="/camera/rgb/image_rect_color" />
  <!-- <arg name="depth_topic"             default="/camera/depth_registered/image_raw" /> -->
  <arg name="camera_info_topic"       default="/camera/rgb/camera_info" />
  <arg name="imu_topic"               default="/imu/data" />
  <arg name="imu_ignore_acc"          default="true" />
  <arg name="imu_remove_gravitational_acceleration" default="true" />
 
 <!-- imu topic -->
   <arg if="$(eval test_type == 'gazebo')" name="imu_raw_topic" default="/imu0"/>
   <arg unless="$(eval test_type == 'gazebo')" name="imu_raw_topic" value="/camera/imu" />

<!-- camera topic -->
   <arg if="$(eval test_type == 'gazebo')" name="left_image_topic" value="/multisense_sl/camera/left/image_raw" />
   <arg unless="$(eval test_type == 'gazebo')" name="left_image_topic" value="/camera/infra1/image_rect_raw" />
   <arg if="$(eval test_type == 'gazebo')" name="left_camera_info_topic" value="/multisense_sl/camera/left/camera_info" />
   <arg unless="$(eval test_type == 'gazebo')" name="left_camera_info_topic" value="/camera/infra1/camera_info" />

   <arg if="$(eval test_type == 'gazebo')" name="right_image_topic" value="/multisense_sl/camera/right/image_raw" />
   <arg unless="$(eval test_type == 'gazebo')" name="right_image_topic" value="/camera/infra2/image_rect_raw" />
   <arg if="$(eval test_type == 'gazebo')" name="right_camera_info_topic" value="/multisense_sl/camera/right/camera_info" />
   <arg unless="$(eval test_type == 'gazebo')" name="right_camera_info_topic" value="/camera/infra2/camera_info" />

<!-- wheel odom topic -->
   <arg if="$(eval test_type == 'gazebo')" name="wheel_odom_topic" default="/wheel_odom"/>
   <arg unless="$(eval test_type == 'gazebo')" name="wheel_odom_topic" value="/odom" />

  <!-- Localization-only mode -->
  <arg name="localization"      default="false"/>
  <arg     if="$(arg localization)" name="rtabmap_args"  default=""/>
  <arg unless="$(arg localization)" name="rtabmap_args"  default="--delete_db_on_start"/>

   <node pkg="imu_complementary_filter" type="complementary_filter_node" name="imu_filter" output="screen">
      <!-- <remap from="imu/data_raw" to="/imu0"/> -->
      <remap from="imu/data_raw" to="$(arg imu_raw_topic)"/>
      <param name="use_mag" value="false"/>
      <param name="world_frame" value="enu"/>
      <param name="publish_tf" value="false"/>
   </node>

  <group ns="rtabmap">
    <!-- Visual Odometry -->
    <node pkg="rtabmap_odom" type="stereo_odometry" name="stereo_odometry" output="$(arg output)" args="$(arg rtabmap_args)">
      <remap from="left/image_rect"        to="$(arg left_image_topic)"/>
      <remap from="right/image_rect"       to="$(arg right_image_topic)"/>
      <remap from="left/camera_info"       to="$(arg left_camera_info_topic)"/>
      <remap from="right/camera_info"      to="$(arg right_camera_info_topic)"/>
      <remap from="odom"                   to="/vo"/>
      <remap from="imu"                    to="$(arg imu_topic)"/>
      <remap from="pose"                   to="$(arg output_pose_topic)"/>

      <param name="output_prefix"          value="$(arg output_prefix)"/>
      <param name="frame_id"               type="string" value="$(arg frame_id)"/>
      <param name="publish_tf"             type="bool"   value="false"/>
      <param name="publish_null_when_lost" type="bool"   value="false"/>
      <param name="guess_frame_id"         type="string" value="odom"/>
      <param name="transform_timeout_odom" value="0.2"/>

      <param name="Odom/FillInfoData"      type="string" value="true"/>
      <param name="Odom/ResetCountdown"    type="string" value="1"/>
      <param name="Vis/FeatureType"        type="string" value="6"/>
      <param name="OdomF2M/MaxSize"        type="string" value="1000"/>
    </node>

    <!-- SLAM -->
    <node name="rtabmap" pkg="rtabmap_slam" type="rtabmap" output="screen" args="$(arg rtabmap_args)">
      <param name="frame_id"         type="string" value="$(arg frame_id)"/>
      <param name="subscribe_depth"  value="false"/>
      <param name="subscribe_stereo" value="true"/>
      <param name="subscribe_rgb"    value="false"/>
      <param name="subscribe_scan"   value="true"/>

      <param name="subscribe_odom_info" value="false"/>
      <param name="approx_sync"         value="false"/>
      <param name="database_path"       value=""/>

      <remap from="left/image_rect"        to="$(arg left_image_topic)"/>
      <remap from="right/image_rect"       to="$(arg right_image_topic)"/>
      <remap from="left/camera_info"       to="$(arg left_camera_info_topic)"/>
      <remap from="right/camera_info"      to="$(arg right_camera_info_topic)"/>

      <param name="map_frame_id"    value="slam_map"/>
      <param name="odom_frame_id"   value=""/>
      <remap from="odom"            to="/odometry/filtered"/>

      <param name="Kp/DetectorStrategy"    type="string" value="6"/> <!-- use same features as odom -->

      <!-- localization mode -->
      <param name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)"/>
      <param     if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
      <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>

    </node>
  </group>

  <node if="$(eval test_type == 'gazebo')" pkg="closedloop_nav_slam" type="wheel_odometry_publisher.py" name="wheel_odometry_publisher" output="screen">
  </node>

  <!-- Odometry fusion (EKF), refer to demo launch file in robot_localization for more info -->
  <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization" clear_params="true" output="screen">

      <param name="frequency" value="50"/>
      <param name="sensor_timeout" value="0.1"/>
      <param name="two_d_mode" value="true"/>

      <param name="odom_frame" value="odom"/>
      <param name="base_link_frame" value="$(arg frame_id)"/>
      <param name="world_frame" value="odom"/>
      <param name="publish_tf"  value="true"/>

      <param name="transform_time_offset" value="0.0"/>

      <param name="odom0" value="/vo"/>
      <param name="odom1" value="$(arg wheel_odom_topic)"/>
      <!-- <param name="imu0" value="$(arg imu_topic)"/> -->

      <!-- The order of the values is x, y, z, roll, pitch, yaw, vx, vy, vz, vroll, vpitch, vyaw, ax, ay, az. -->
      <rosparam param="odom0_config">[true, true, true,
                                      true, true, true,
                                      false, false, false,
                                      false, false, false,
                                      false, false, false]</rosparam>

      <rosparam param="odom1_config">[true, true, true,
                                      true, true, true,
                                      false, false, false,
                                      false, false, false,
                                      false, false, false]</rosparam>

      <!-- <rosparam     if="$(arg imu_ignore_acc)" param="imu0_config">[
                                     false, false, false,
                                     true,  true,  true,
                                     false, false, false,
                                     true,  true,  true,
                                     false,  false,  false] </rosparam>
      <rosparam unless="$(arg imu_ignore_acc)" param="imu0_config">[
                                     false, false, false,
                                     true,  true,  true,
                                     false, false, false,
                                     true,  true,  true,
                                     true,  true,  true] </rosparam> -->

      <param name="odom0_differential" value="false"/>
      <param name="odom1_differential" value="false"/>
      <!-- <param name="imu0_differential" value="false"/> -->

      <param name="odom0_relative" value="true"/>
      <param name="odom1_relative" value="true"/>
      <!-- <param name="imu0_relative" value="true"/> -->

      <!-- <param name="imu0_remove_gravitational_acceleration" value="$(arg imu_remove_gravitational_acceleration)"/> -->

      <param name="print_diagnostics" value="true"/>

      <!-- ======== ADVANCED PARAMETERS ======== -->
      <param name="odom0_queue_size" value="5"/>
      <param name="odom1_queue_size" value="5"/>
      <!-- <param name="imu0_queue_size" value="50"/> -->

      <!-- The values are ordered as x, y, z, roll, pitch, yaw, vx, vy, vz,
           vroll, vpitch, vyaw, ax, ay, az. -->
      <rosparam param="process_noise_covariance">[0.005, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0.005, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0.006, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0.003, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0.003, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0.006, 0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0.0025, 0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0.0025, 0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0.004, 0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0.001, 0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0.001, 0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0.002, 0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0.001, 0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0.001, 0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0.0015]</rosparam>

      <!-- The values are ordered as x, y,
           z, roll, pitch, yaw, vx, vy, vz, vroll, vpitch, vyaw, ax, ay, az. -->
           <rosparam param="initial_estimate_covariance">[1e-9, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    1e-9, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    1e-9, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    1e-9, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    1e-9, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    1e-9, 0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    1e-9, 0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    1e-9, 0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    1e-9, 0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-9,  0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-9,  0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-9,  0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-9, 0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-9, 0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-9]</rosparam>

    </node>
</launch>
